### ISSUE TITLE
buildkit scheduler error: return leaving incoming open

### ISSUE THREAD

**Issue by jmgilman** [earthly/earthly#2957]
We've been seeing the following error mostly in CI, however, we've also seen it on a developer's machine as well. The issue reproduced on the developer machine several times and only went away after changing the base image of the target (seems sort of random, but perhaps the team can make more sense of it). 

```
Error: build target: build main: failed to solve: unlazy force execution: buildkit scheduler error: return leaving incoming open. Please report this with BUILDKIT_SCHEDULER_DEBUG=1
```

On CI, the job often fails multiple times in a row if you just re-run it as is. I haven't really been able to make sense of why it comes and goes yet.

EDIT by the Earthly team: To get unstuck in this situation you can restart buildkit with `docker stop earthly-buildkitd`.

EDIT 2 by the Earthly team: Possibly related issue: https://github.com/earthly/earthly/issues/2454 (inconsistent graph state error).

EDIT 3 by the Earthly team: This is now a top priority for us. We will provide weekly updates on this issue until it is resolved.

**Comment by ezeev** [comment]
I was able to "reproduce" this on my linux machine while building [earthly-solutions](https://github.com/earthly/earthly-solutions). Reproduce in quotes because it seems to happen at random places.

earthly version v0.7.6 78d2f5be89c8ada1048ed7130ed319948bf79912 linux/amd64; Ubuntu 22.04.2 LTS (Jammy Jellyfish)

**Comment by vladaionescu** [comment]
> Restarting Docker on my machine seems to have fixed it.

I believe restarting buildkit should be enough to get it back to a working state `docker stop earthly-buildkitd`.

**Comment by vladaionescu** [comment]
This issue is now high-priority for the team, as it's been picking up in frequency lately quite a bit.

**Comment by xsteadfastx** [comment]
for us too.

**Comment by GMaxera** [comment]
We are seeing this on our CI too, but docker stop is not a viable solution for us because we are using the earthly remote builder approach.
So, we need a fix or another workaround to avoid that.

**Comment by alexcb** [comment]
building off https://github.com/earthly/earthly/issues/2957#issuecomment-1563621287

I was able to reproduce it in roughly 20 minutes a few times now, using this script:

```
#!/bin/sh
set -e
while true; do
  date
  cat go_server/main.go | sed "s/Hello .* Earthly/Hello $(date) Earthly/g" > go_server/main.go_ && mv go_server/main.go_ go_server/main.go
  cat python_server/src/quote.py | sed "s/Hello .* Earthly/Hello $(date) Earthly/g" > python_server/src/quote.py_ && mv python_server/src/quote.py_ python_server/src/quote.py
  earthly-v0.7.6 -V -P +main-pipeline >log1 2>&1 >log3 &
  earthly-v0.7.6 -V -P +main-pipeline >log2 2>&1 >log3 &
  earthly-v0.7.6 -V -P +main-pipeline >log3 2>&1 >log3 &
  wait

  if grep BUILDKIT_SCHEDULER_DEBUG log1; then
    exit 1
  fi
  if grep BUILDKIT_SCHEDULER_DEBUG log2; then
    exit 1
  fi
  if grep BUILDKIT_SCHEDULER_DEBUG log3; then
    exit 1
  fi
done
```

**Comment by alexcb** [comment]
this bug still persists in 367e294d080db21b89f50d2a2b8c5c496e851a44 (which includes upstream buildkit changes from 2023-05-30)

**Comment by alexcb** [comment]
this bug persists in v0.7.8; it was not related to the "unknown blob in history" error.

**Comment by mitre-tm** [comment]
This bug persists in v0.7.9. It happened for me on Earthly Satellite
```
*  moon    linux/amd64  xlarge  v0.7.9   operational
```

**Comment by vladaionescu** [comment]
This is a top priority. It seems to be a buildkit upstream bug, and we have a reproduction case (although it's rather complex and non deterministic). It's hard to come up with an estimate on when this is fixed, but hopefully soon.

**Comment by alexcb** [comment]
We have discovered some data race warnings inside buildkit. Here's some source locations, for example:

```
$ docker logs earthly-dev-buildkitd 2>&1 | grep -A 2 'WARNING: DATA RACE' | grep 'github\.com/moby' | sort | uniq
  github.com/moby/buildkit/cache/contenthash.(*cacheManager).GetCacheContext()
  github.com/moby/buildkit/executor/resources.(*Sampler[...]).run()
  github.com/moby/buildkit/executor/resources.(*Sub[...]).Close()
  github.com/moby/buildkit/executor/resources/types.(*Sample).Timestamp()
  github.com/moby/buildkit/executor/resources/types.(*SysCPUStat).MarshalJSON()
  github.com/moby/buildkit/executor/resources/types.(*SysSample).Timestamp()
  github.com/moby/buildkit/solver.(*Job).Build()
  github.com/moby/buildkit/solver.(*Job).RegisterCompleteTime()
  github.com/moby/buildkit/solver.(*Solver).connectProgressFromState()
  github.com/moby/buildkit/util/network/cniprovider.(*cniNS).Close()
  github.com/moby/buildkit/util/network/cniprovider.(*cniNS).sample()
  github.com/moby/buildkit/util/progress.(*Progress).Meta()
  github.com/moby/buildkit/util/progress.(*progressWriter).WriteRawProgress()
  github.com/moby/buildkit/util/pull/pullprogress.trackProgress()
```

In order to produce these warnings, we had to manually enable the `-race` flag during compilation of `buildkitd` -- this has prompted https://github.com/moby/buildkit/pull/3994 to make it easier for both us (and the larger buildkit community) to debug these race conditions.

It's not clear if these data races are related to this `buildkit scheduler error` but it's a promising lead (and will at the very least increase code quality).

**Comment by alexcb** [comment]
We've gotten 3 data race PRs merged into buildkit:
- https://github.com/moby/buildkit/pull/4004 
- https://github.com/moby/buildkit/pull/4008
- https://github.com/moby/buildkit/pull/4010

We're continuing to work through other warnings.

**Comment by alexcb** [comment]
A quick update: We're still actively working on this and have been pulling in upstream changes into our buildkit fork and have run into some cgroup-related yak shaving.

**Comment by alexcb** [comment]
we are still looking into this issue -- it is difficult to debug.

**Comment by alexcb** [comment]
I found a reproduction, and it's embarrassingly simple!

```Earthfile
VERSION 0.7

foo:
  FROM alpine:3.17
  COPY my-file .
  RUN true
```

Previously we thought this race condition was related to high-cpu usage, but it's not -- **it's a combination of a COPY and the wait-end feature**.

It can be reproduced fairly consistently with:

```
$ date > my-file; earthly-v0.7.19 +foo & earthly-v0.7.19 +foo
[1] 1261981
 Init 🚀
————————————————————————————————————————————————————————————————————————————————

 Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Starting buildkit daemon as a docker container (earthly-buildkitd)...
           buildkitd | ...Done
           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)

Streaming logs to https://cloud.earthly.dev/builds/e6ce6c80-0d56-4eb1-a86d-e6de171732e8

 Build 🔧
————————————————————————————————————————————————————————————————————————————————

         alpine:3.17 | --> Load metadata alpine:3.17 linux/amd64

Streaming logs to https://cloud.earthly.dev/builds/3fb6298f-a6eb-4bbc-a687-4293ae62050f

 Build 🔧
————————————————————————————————————————————————————————————————————————————————

         alpine:3.17 | --> Load metadata alpine:3.17 linux/amd64
                +foo | --> FROM alpine:3.17
--> FROM alpine:3.17
[----------] 100% FROM alpine:3.17
                +foo | [----------] 100% FROM alpine:3.17
                +foo | --> COPY my-file .
                +foo | --> RUN true
              output | --> exporting outputs
View logs at https://cloud.earthly.dev/builds/e6ce6c80-0d56-4eb1-a86d-e6de171732e8
            _unknown *failed* | build target: build main: failed to solve: buildkit scheduler error: return leaving incoming open. Please report this with BUILDKIT_SCHEDULER_DEBUG=1
Error: build target: build main: failed to solve: buildkit scheduler error: return leaving incoming open. Please report this with BUILDKIT_SCHEDULER_DEBUG=1
```

I have tested a similar Dockerfile with `buildx`, which doesn't contain the bug.

It appears that this bug was introduced with the `--wait-block` feature (even if the `WAIT` / `END` commands are not used).

If the Earthfile's version is changed to:

```
VERSION \
  --check-duplicate-images \
  --earthly-git-author-args \
  --earthly-locally-arg \
  --earthly-version-arg \
  --explicit-global \
  --new-platform \
  --no-tar-build-output \
  --save-artifact-keep-own \
  --shell-out-anywhere \
  --use-cache-command \
  --use-chmod \
  --use-copy-link \
  --use-host-command \
  --use-no-manifest-list \
  --use-pipelines \
  --use-project-secrets \
  #--wait-block \
  0.6
```

which is equivalent to VERSION 0.7 except for the wait-block feature, running multiple copies of earthly doesn't appear to trigger the bug.

We will look more closely into how the wait block feature changes the buildkit scheduler code execution path.

In the meanwhile, if you are not using any WAIT/END blocks, you may want to use the above `VERSION ..... 0.6` definition instead.

**Comment by alexcb** [comment]
This issue exists upstream.

```
alex@surfperch:~/gh/moby/buildkit$ rm -rf bin/build
alex@surfperch:~/gh/moby/buildkit$ git rev-parse HEAD
77b9e116cc337499e98090f407d6695fd73c8201
alex@surfperch:~/gh/moby/buildkit$ make binaries 
docker buildx bake binaries
[+] Building 1.4s (30/30) FINISHED                                                                                                                                                                  docker:default
 => [internal] load build definition from Dockerfile                                                                                                                                                          0.0s
 => => transferring dockerfile: 13.35kB                                                                                                                                                                       0.0s
 => [internal] load .dockerignore                                                                                                                                                                             0.0s
 => => transferring context: 56B                                                                                                                                                                              0.0s
 => resolve image config for docker.io/docker/dockerfile-upstream:master                                                                                                                                      0.4s
 => CACHED docker-image://docker.io/docker/dockerfile-upstream:master@sha256:b08a31f8f58d1998b8d4737957cb7f581c08972f550878126360daa90181042f                                                                 0.0s
 => [internal] load metadata for docker.io/library/alpine:3.18                                                                                                                                                0.4s
 => [internal] load metadata for docker.io/tonistiigi/xx:1.2.1                                                                                                                                                0.4s
 => [internal] load metadata for docker.io/tonistiigi/binfmt:buildkit-v7.1.0-30@sha256:45dd57b4ba2f24e2354f71f1e4e51f073cb7a28fd848ce6f5f2a7701142a6bf0                                                       0.0s
 => [internal] load metadata for docker.io/library/golang:1.20-alpine3.18                                                                                                                                     0.4s
 => [internal] load build context                                                                                                                                                                             0.1s
 => => transferring context: 543.72kB                                                                                                                                                                         0.1s
 => [xx 1/1] FROM docker.io/tonistiigi/xx:1.2.1@sha256:8879a398dedf0aadaacfbd332b29ff2f84bc39ae6d4e9c0a1109db27ac5ba012                                                                                       0.0s
 => [golatest 1/1] FROM docker.io/library/golang:1.20-alpine3.18@sha256:c63dbdb3cca37abbee4c50f61e34b1d043c2669d03f34485f9ee6fe5feed4e48                                                                      0.0s
 => FROM docker.io/tonistiigi/binfmt:buildkit-v7.1.0-30@sha256:45dd57b4ba2f24e2354f71f1e4e51f073cb7a28fd848ce6f5f2a7701142a6bf0                                                                               0.0s
 => [git 1/2] FROM docker.io/library/alpine:3.18@sha256:7144f7bab3d4c2648d7e59409f15ec52a18006a128c733fcff20d3a4a54ba44a                                                                                      0.0s
 => CACHED [gobuild-base 1/2] RUN apk add --no-cache file bash clang lld musl-dev pkgconfig git make                                                                                                          0.0s
 => CACHED [gobuild-base 2/2] COPY --link --from=xx / /                                                                                                                                                       0.0s
 => CACHED [runc 1/3] WORKDIR /go/src/github.com/opencontainers/runc                                                                                                                                          0.0s
 => CACHED [runc 2/3] RUN set -e; xx-apk add musl-dev gcc libseccomp-dev libseccomp-static;   [ "$(xx-info arch)" != "ppc64le" ] || XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple                     0.0s
 => CACHED [git 2/2] RUN apk add --no-cache git                                                                                                                                                               0.0s
 => CACHED [runc-src 1/2] WORKDIR /usr/src                                                                                                                                                                    0.0s
 => CACHED [runc-src 2/2] RUN git clone https://github.com/opencontainers/runc.git runc   && cd runc && git checkout -q "v1.1.9"                                                                              0.0s
 => CACHED [runc 3/3] RUN --mount=from=runc-src,src=/usr/src/runc,target=. --mount=target=/root/.cache,type=cache   CGO_ENABLED=1 xx-go build -mod=vendor -ldflags '-extldflags -static' -tags 'apparmor sec  0.0s
 => CACHED [binaries-linux 1/4] COPY --link --from=runc /usr/bin/runc /buildkit-runc                                                                                                                          0.0s
 => CACHED [binaries-linux 2/4] COPY --link --from=tonistiigi/binfmt:buildkit-v7.1.0-30@sha256:45dd57b4ba2f24e2354f71f1e4e51f073cb7a28fd848ce6f5f2a7701142a6bf0 / /                                           0.0s
 => CACHED [buildkit-base 1/1] WORKDIR /src                                                                                                                                                                   0.0s
 => CACHED [buildkit-version 1/1] RUN --mount=target=.   PKG=github.com/moby/buildkit VERSION=$(git describe --match 'v[0-9]*' --dirty='.m' --always --tags) REVISION=$(git rev-parse HEAD)$(if ! git diff -  0.0s
 => CACHED [buildctl 1/1] RUN --mount=target=. --mount=target=/root/.cache,type=cache   --mount=target=/go/pkg/mod,type=cache   --mount=source=/tmp/.ldflags,target=/tmp/.ldflags,from=buildkit-version   xx  0.0s
 => CACHED [binaries-linux 3/4] COPY --link --from=buildctl /usr/bin/buildctl /                                                                                                                               0.0s
 => CACHED [buildkitd 1/1] RUN --mount=target=. --mount=target=/root/.cache,type=cache   --mount=target=/go/pkg/mod,type=cache   --mount=source=/tmp/.ldflags,target=/tmp/.ldflags,from=buildkit-version   x  0.0s
 => CACHED [binaries-linux 4/4] COPY --link --from=buildkitd /usr/bin/buildkitd /                                                                                                                             0.0s
 => exporting to client directory                                                                                                                                                                             0.3s
 => => copying files 147.28MB                                                                                                                                                                                 0.2s
alex@surfperch:~/gh/moby/buildkit$ sudo bin/build/buildkitd
INFO[2023-09-25T16:39:47-07:00] auto snapshotter: using overlayfs            
WARN[2023-09-25T16:39:47-07:00] using host network as the default            
INFO[2023-09-25T16:39:48-07:00] found worker "z46piheuvr1cv8n524eys5fi9", labels=map[org.mobyproject.buildkit.worker.executor:oci org.mobyproject.buildkit.worker.hostname:surfperch org.mobyproject.buildkit.worker.network:host org.mobyproject.buildkit.worker.oci.process-mode:sandbox org.mobyproject.buildkit.worker.selinux.enabled:false org.mobyproject.buildkit.worker.snapshotter:overlayfs], platforms=[linux/amd64 linux/amd64/v2 linux/amd64/v3 linux/arm64 linux/riscv64 linux/ppc64 linux/ppc64le linux/s390x linux/386 linux/mips64le linux/mips64 linux/arm/v7 linux/arm/v6] 
WARN[2023-09-25T16:39:48-07:00] using host network as the default            
INFO[2023-09-25T16:39:48-07:00] found worker "v87ci4d74k1abtg0upscop067", labels=map[org.mobyproject.buildkit.worker.containerd.namespace:buildkit org.mobyproject.buildkit.worker.containerd.uuid:4b1aff2e-2150-4bf2-8877-3ea25113ec86 org.mobyproject.buildkit.worker.executor:containerd org.mobyproject.buildkit.worker.hostname:surfperch org.mobyproject.buildkit.worker.network:host org.mobyproject.buildkit.worker.selinux.enabled:false org.mobyproject.buildkit.worker.snapshotter:overlayfs], platforms=[linux/amd64 linux/amd64/v2 linux/amd64/v3 linux/arm64 linux/riscv64 linux/ppc64 linux/ppc64le linux/s390x linux/386 linux/mips64le linux/mips64 linux/arm/v7 linux/arm/v6] 
INFO[2023-09-25T16:39:48-07:00] found 2 workers, default="z46piheuvr1cv8n524eys5fi9" 
WARN[2023-09-25T16:39:48-07:00] currently, only the default worker can be used. 
INFO[2023-09-25T16:39:48-07:00] running server on /run/buildkit/buildkitd.sock 
INFO[2023-09-25T16:39:58-07:00] detected 127.0.0.53 nameserver, assuming systemd-resolved, so using resolv.conf: /run/systemd/resolve/resolv.conf 
ERRO[2023-09-25T16:39:58-07:00] buildkit scheluding error: leaving incoming open. forcing solve. Please report this with BUILDKIT_SCHEDULER_DEBUG=1 
ERRO[2023-09-25T16:39:59-07:00] /moby.buildkit.v1.Control/Solve returned error: rpc error: code = Unknown desc = buildkit scheduler error: return leaving incoming open. Please report this with BUILDKIT_SCHEDULER_DEBUG=1 
```

and here's an reproduction program that talks to buildkit directly (i.e. no earthly code):

```
$ cat main.go 
package main

import (
	"context"
	"fmt"
	"os"
	"sync"
	"time"

	"github.com/containerd/containerd/platforms"
	"github.com/moby/buildkit/client"
	"github.com/moby/buildkit/client/llb"
	gwclient "github.com/moby/buildkit/frontend/gateway/client"
	"golang.org/x/sync/errgroup"
)

// repro case: go build main.go && echo "the date is $(date)" > src.txt && sudo /bin/sh -c 'export BUILDKIT_ADDR=unix:///run/buildkit/buildkitd.sock; ./main & ./main & ./main & ./main& ./main& ./main& ./main& ./main& ./main& ./main & wait' 2>&1

// buildkitAddr allows a user to export BUILDKIT_ADDR=unix:///run/buildkit/buildkitd.sock which will connect to a locally runing buildkitd binary.
func buildkitAddr() string {
	buildkitAddr := os.Getenv("BUILDKIT_ADDR")
	if buildkitAddr == "" {
		buildkitAddr = "tcp://127.0.0.1:8372"
	}
	return buildkitAddr
}

func main() {
	bkClient, err := client.New(context.TODO(), buildkitAddr())
	if err != nil {
		fmt.Printf("failed to connect to buildkit: %v\n", err)
		os.Exit(1)
	}
	defer bkClient.Close()

	ch := make(chan *client.SolveStatus)
	go logStatus(ch)

	localPath, err := os.Getwd()
	if err != nil {
		fmt.Printf("failed to current path: %v\n", err)
		os.Exit(1)
	}

	// bkClient.Build and the buildFunc is particularly useful when you have multiple LLB states
	// and you want buildkit to solve them in parallel
	buildFunc := func(ctx context.Context, gwClient gwclient.Client) (*gwclient.Result, error) {
		src := llb.Local("cwd")

		state := llb.Image("alpine:latest").
			File(llb.Copy(src, "src.txt", "/dest")).
			Run(llb.Shlex("cat /dest")).Root()

		platform := "linux/amd64"

		var gmu sync.Mutex

		eg, solveCtx := errgroup.WithContext(context.Background()) // keep this open after bf finishes
		for i := 0; i < 2; i++ {
			eg.Go(func() error {
				state = state.Run(llb.Shlex("true")).Root()
				gmu.Lock()
				def, err := state.Marshal(ctx, llb.Platform(platforms.MustParse(platform)))
				gmu.Unlock()
				if err != nil {
					return err
				}

				r, err := gwClient.Solve(solveCtx, gwclient.SolveRequest{
					Definition: def.ToPB(),
				})
				if err != nil {
					return err
				}

				ref, err := r.SingleRef()
				if err != nil {
					return err
				}
				_, err = ref.ReadDir(ctx, gwclient.ReadDirRequest{Path: "/"})
				return err
			})
		}
		defer func() {
			err := eg.Wait()
			if err != nil {
				panic(err)
			}
		}()

		state = state.Run(llb.Shlex("true")).Root()
		gmu.Lock()
		def, err := state.Marshal(ctx, llb.Platform(platforms.MustParse(platform)))
		gmu.Unlock()
		if err != nil {
			return nil, err
		}

		time.Sleep(time.Second)

		r, err := gwClient.Solve(solveCtx, gwclient.SolveRequest{
			Definition: def.ToPB(),
		})
		if err != nil {
			return nil, err
		}

		ref, err := r.SingleRef()
		if err != nil {
			return nil, err
		}

		// multiple references can be returned together with buildFunc
		result := gwclient.NewResult()
		result.AddRef("state", ref)
		return result, nil
	}

	solveOpt := client.SolveOpt{
		LocalDirs: map[string]string{
			"cwd": localPath,
		},
	}

	resp, err := bkClient.Build(context.TODO(), solveOpt, "", buildFunc, ch)
	if err != nil {
		panic(err)
		//fmt.Printf("failed to build llb: %v\n", err)
		//os.Exit(1)
	}

	fmt.Printf("Solve Response: %v\n", resp)
}

func logStatus(ch chan *client.SolveStatus) {
	for {
		status := <-ch
		if status == nil {
			break
		}
		fmt.Printf("status is %v\n", status)
		for _, v := range status.Vertexes {
			fmt.Printf("====vertex: %+v\n", v)
		}
		for _, s := range status.Statuses {
			fmt.Printf("====status: %+v\n", s)
		}
		for _, l := range status.Logs {
			fmt.Printf("====log: %s\n", string(l.Data))
		}
	}
}
```

```
$ go build main.go && echo "the date is $(date)" > src.txt && echo lets run it && sudo /bin/sh -c 'export BUILDKIT_ADDR=unix:///run/buildkit/buildkitd.sock && for i in $(seq 1 2); do ./main & done && wait' 2>&1 | grep panic
lets run it
panic: failed to solve: buildkit scheduler error: return leaving incoming open. Please report this with BUILDKIT_SCHEDULER_DEBUG=1
```

Note that the call to `time.Sleep(time.Second)` increases the likelihood of the error (without it, it doesn't occur frequently).

**Comment by alexcb** [comment]
upstream issue: https://github.com/moby/buildkit/issues/4278

**Comment by alexcb** [comment]
I tried deduplicating calls to `forceExecution` as a work-around under #3300, but it failed.

**Comment by alexcb** [comment]
we're currently working on pulling upstream changes into our fork, so when the upstream issue is resolved it'll be easier to pull it in.

**Comment by alexcb** [comment]
The initial upstream merge was done in d919ffa6465b0f5ac09769e4efa8b6e077ab5990; however the upstream fix under buildkit is still being worked on. When it gets merged, we'll be able to pull it in more easily since we're much closer to moby/buildkit:master now.

**Comment by mwopfner** [comment]
@alexcb That is good to hear that you can now more easily pull upstream fixes/changes. Is there any plan to get the remaining changes upstream or are they just to specific to earthly?

**Comment by alexcb** [comment]
This has been released as of [v0.7.21](https://github.com/earthly/earthly/releases/tag/v0.7.21); please let us know if this occurs again.

**Comment by vladaionescu** [comment]
Just a quick update: Earthly Satellites now also contain this update. If auto-updates are enabled (they are by default), you should be getting this overnight.
